File: dagger2/src/main/java/pgreze/dagger2/app/AppComponent.java
package pgreze.dagger2.app;

import android.content.Context;

import javax.inject.Singleton;

import dagger.Component;
import pgreze.dagger2.analytics.AnalyticsManager;

@Singleton
@Component(modules = {
        AppModule.class
})
public interface AppComponent {
    void inject(Dagger2App app);

    Context context();
    AnalyticsManager analyticsManager();
}

File: dagger/android/src/main/kotlin/pgreze/dagger2/app/AppComponent.kt
package pgreze.dagger2.app

import android.content.Context

import javax.inject.Singleton

import dagger.Component
import pgreze.dagger2.analytics.AnalyticsManager

/**
 * App component.
 * With @Singleton annotation, all @Singleton objects will be global to the app.
 * @see pgreze.dagger2.di.AndroidLifecycleScope for activity/service single and lifecycle related
 */
@Singleton
@Component(modules = arrayOf(AppModule::class))
interface AppComponent {
    fun inject(app: Dagger2App)

    // We have to manually defined these accessors for objects used by other components

    fun context(): Context
    fun analyticsManager(): AnalyticsManager
}
